src/body.js

```diff
@@ -47,16 +61,71 @@ class Body {
   }
 
   async text () {
-    if (this[kBody] == null) return ''
+    if (this[kBody] === null) return ''
+
+    return (await consumeBody(this[kBody])).toString('utf-8')
+  }
+}
 
-    this[kBody].setEncoding('utf8')
-    let res = ''
-    for await (const chunk of this[kBody]) {
-      if (!this[kBodyUsed]) this[kBodyUsed] = true
-      res += chunk
+async function consumeBody (body) {
+  if (isUnusable(body)) throw TypeError('cannot consume unusable body')
+
+  const bufs = []
+  for await (const chunk of body) {
+    bufs.push(Buffer.from(chunk))
+  }
+
+  return Buffer.concat(bufs)
+}
+
+function isUnusable (body) {
+  return body?.disturbed ?? false
+}
+
+function createAsyncIterableFromBlob (blob) {
+  return {
+    async * [Symbol.asyncIterator] () {
+      yield * await blob.text()
     }
-    return res
   }
 }
 
-module.exports = Body
+function extractBody (body, keepalive = false) {
+  // Test for unique iterator types (URLSearchParams, String, or ArrayBuffer) before the isAsyncIterable check
+
+  // todo: handle FormBody
+
+  if (body instanceof URLSearchParams) {
+    // spec says to run application/x-www-form-urlencoded on body.list
+    // this is implemented in Node.js as apart of an URLSearchParams instance toString method
+    // See: https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/url.js#L490
+    // And: https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/url.js#L1100
+    return [
+      body.toString(),
+      'application/x-www-form-urlencoded;charset=UTF-8'
+    ]
+  } else if (typeof body === 'string') {
+    return [
+      Buffer.from(body, 'utf-8').toString(),
+      'text/plain;charset=UTF-8'
+    ]
+  } else if (body instanceof ArrayBuffer || ArrayBuffer.isView(body)) {
+    return [Buffer.from(body), null]
```

Comment:
Can we avoid using `Buffer`? That's node specific.

Reply:
Yeah; i'm struggling to understand how to use ArrayBuffer. Would a DataView be better?

Reply:
My general idea is to make any valid body input into some type of iterable. Buffer seems like the most efficient way to go from string -> bytes. I think I could then pass these buffer instances to something like `Uint8Array` and then use that instead?

Reply:
Alternatively i found something like this: https://coolaj86.com/articles/unicode-string-to-a-utf-8-typed-array-buffer-in-javascript/ which is not node-specific

Reply:
I'm sticking with Buffer for now since this is going to be a node specific library. Uint8Array may be a good next step if it comes with better memory/performance improvements as well